selfhost/typechecker.jakt

```diff
@@ -6061,148 +5991,296 @@ struct Typechecker {
+                                found_optional = false
+                                yield None
+                            }
+                        }
+                    }
+                    yield Some(struct_id ?? StructOrEnumId::Struct(optional_struct_id))
+                }
+                else => Some(StructOrEnumId::Struct(id))
             }
+            else => {
+                .error(format("no methods available on value (type: {})", .type_name(checked_expr_type_id)), checked_expr.span())
+                yield None
+            }
+        }
 
-            return match result! {
-                Return(x) => value_to_checked_expression(x, interpreter)
-                Throw(x) => {
-                    .error(
-                        format("Compiletime call failed: {}", x)
-                        span
-                    )
-                    yield checked_call
+        if is_optional and not found_optional {
+            .error(format("Optional chain mismatch: expected optional chain, found {}", .type_name(checked_expr_type_id)), checked_expr.span())
+        }
+
+        let function_id = match parent_id.has_value() {
+            true => .resolve_method_call(parent_id: parent_id!, call, span)
+            // skip any resolution since we already gave an error explaining why
+            // we cannot resolve a method here.
+            else => None
+        }
+
+        guard function_id.has_value() else {
+            return .checked_call_to_method_call(
+                expr: checked_expr
+                checked_call: .typecheck_unresolved_call(call, scope_id, safety_mode)
+                is_optional, span)
+        }
+
+        let callee = .get_function(function_id!)
+
+        let checked_call = .typecheck_resolved_call(
+            call, caller_scope_id: scope_id, span
+            function_id: function_id!, safety_mode,
+            resolved_namespaces: [],
+            this_expr_type_id: match callee.is_static() {
+                false => Some(checked_expr_type_id)
+                else => {
+                    .error("Cannot call static method on an instance of an object", span)
+                    yield None
                 }
             }
+            type_hint)
+
+        if callee.is_mutating() and not checked_expr.is_mutable(program: .program) {
+            .error("Cannot call mutating method on an immutable object instance", span)
+        }
+
+        return .execute_call_if_comptime(
+            checked_call, span,
+            caller_scope_id: scope_id
+            resolved_namespaces: []
+            this_expr: checked_expr
+        ) ?? .checked_call_to_method_call(expr: checked_expr, checked_call, is_optional, span)
+    }
+
+    function checked_call_to_method_call(mut this, expr: CheckedExpression, checked_call: CheckedCall, is_optional: bool, span: Span) throws -> CheckedExpression =>
```

Comment:
Arguably for readability multiline arrow function like this would be better as a regular function.